### ISSUE TITLE
PortAudio ASIO driver scan

### ISSUE THREAD

**Issue by jjYBdx4IL** [PortAudio/portaudio#696]
Looks like ASIO drivers aren't meant to be loaded automatically. Please don't automatically load the drivers. Every ASIO app I know loads the driver only when the user selects it. Or at least add an option to disable that.

**Comment by jjYBdx4IL** [comment]
Also, there seems to be no functionality to open the ASIO driver settings/configuration dialog. The standard seems to be (from what I know by using DAWs under Windows):

- user selects from ASIO driver list
- driver gets loaded when user selects it
- DAW UI has button to access ASIO driver settings page (provided by the specific ASIO driver itself)

So, PortAudio would also need a way to call that config dialog and a way to retrieve the ASIO driver list.

**Comment by RossBencina** [comment]
I understand your request. However what you're suggesting is to change the global PortAudio device enumeration behavior that has been part of PortAudio since 2005 or earlier. It's not going to change any time soon.

PortAudio enumerates all devices when `Pa_Initialize` is called. We considered enumerating devices lazily but this causes other issues such as non-deterministic crashing. We decided it's better to have deterministic crashing at the time that `Pa_Initialize` is called. This is all ancient history now and is unlikely to change any time soon. 

The function to display the ASIO settings dialog is declared in `pa_asio.h`. Note that it has limitations (you can't show settings for device A while device B is open).

**Comment by jjYBdx4IL** [comment]
How about handling ASIO independently from the generic API? Would probably still be better than building PortAudio without ASIO support, and then adding ASIO support separately in every program just to have proper support for it.

**Comment by dechamps** [comment]
FWIW I believe there should be very big warnings/documentation when building PortAudio with ASIO support, on the grounds that just having a faulty ASIO driver installed anywhere on the system can make the *entire* application process crash as soon as it tries to initialize PortAudio, *even if the user did not intend to use any ASIO devices at all*. I suspect many people building PortAudio are not aware that PortAudio's ASIO support is a giant footgun in that regard. The fact that this is just a consequence of PortAudio's design doesn't make it any less dangerous.

Basically, if you're not okay with loading potentially unbounded amounts of potentially crash-happy third-party code from random sources (a lot of which are not known for writing high-quality, reliable code) at initialization time, then don't enable ASIO support when building PortAudio. That's really asking for trouble.

**Comment by jmelas** [comment]
Apart from crashing, enumerating ASIO devices when the application starts adds significant delay in the startup-time of the Application.

**Comment by oifii** [comment]
Since I switched to portaudio version 19, built enabling all audio API defines, including the asio one, every time I launch any of my many portaudio-based apps , the global windows 10 audio shuts for a second or two (while Pa_Initialize(); is being called. The audio streaming from youtube/chrome browser on a USB device (id=3, hostapi=MME, devicename=Line (USB AUDIO  CODEC) is interrupted for a second or two. On my windows 10 system, I do have one asio audio device (id=26, hostapi=ASIO, devicename=Realtek ASIO) that is a dummy asio device. On this same system with the same connected audio devices, this situation was not occurring when launching my portaudio -based audio apps relying on previous portaudio library version (i.e. source code snapshot from before 2016). Is it portaudio asio code that has changed? Must a more recent asio SDK be used? more recent than the 2.0 asio SDK? Or is it windows that changed something around the audio drivers access mechanism/behavior?

**Comment by syntaro** [comment]
hello there.

Im from japan.
And oifii, Im having same problem. 
My note pc have RealTekASIO installed.
Cant remove. and When I uninstall ASIO4All it semms no problem.

#904 have code that skip open and get detail. 
so no detail in app life time.

but My PC work fine . My PC RealTek +ASIO4ALL both installed.


Just missing samplerate etc.And it work.

I Just comment out 3 lines. Its all from me.

Could it help something?

**Comment by mirh** [comment]
> I suspect many people building PortAudio are not aware that PortAudio's ASIO support is a giant footgun in that regard.

You can't just build it by accident, when you have to provide yourself the SDK?

> Basically, if you're not okay with loading potentially unbounded amounts of potentially crash-happy third-party code from random sources at initialization time

I mean, that's just all device drivers ever though, isn't it? Though admittedly the ASIO drivers quality is more of a mixed bag. 
But if any, the real universal trouble is the startup delay (for as much as that seems to have been far smaller in the past?)

How about just adding some kind of runtime knob, that then programs can just ship disabled by default to avoid querying even though the api is compiled in? Syntaro's `INIT_NOT_DETAIL` hack in #904 already seeming kind of this.

**Comment by mirh** [comment]
> Well if you use CMake, [it's downloaded for you automatically](https://github.com/PortAudio/portaudio/blob/52d50e2f4d2f3ab815f64fe683d864d99011186d/CMakeLists.txt#L187).

Lool. Touché (even though it's disabled by default). 

> An unsuspecting PortAudio user would not necessarily expect that just enabling ASIO at build time

I don't think "users" are the sort of people that randomly plays around with compile flags..

I do agree on the possible unsuspectingness though, for the average developer that is at least. 

> which is much harder to do if drivers are being loaded silently in the background like PortAudio is doing at init time.

Right.. So I suppose here the bigger problem is kinda the lack of lazy loading...
Or perhaps, maybe I get portaudio's point in having/pre-loading a list of all the available devices. But couldn't it be possible to just query/list the *names* of the registered ASIO drivers, without actually even loading their dlls?

> only call into purely OS-provided APIs that behave in reliable and predictable ways - they do not load third-party code.

I was actually thinking that likely biggest difference (aside of stability, amateurishness, or whatever) - is that for the OS native apis it would always have itself a cached list that it can then immediately return to applications (and if there were to be any problems with something, they would show even on the desktop with nothing else opened). 
With asio instead (unless perhaps having asiomulti in-between?) you are effectively also initializing the driver.

**Comment by dechamps** [comment]
> I don't think "users" are the sort of people that randomly plays around with compile flags..

To be clear, by "users" I mean users of PortAudio, i.e. application developers. I use the term "end user" to refer to the users of the apps.

> But couldn't it be possible to just query/list the names of the registered ASIO drivers, without actually even loading their dlls?

Oh that is absolutely possible, but the problem is, that's not enough. Due to the way the PortAudio API works, it's not just the names you need to know at initialization time - you need to know things like default sample rates and maximum channel counts as well, and *that* definitely requires loading the driver. That is the root cause of the problem and a fundamental limitation of PortAudio that cannot easily be fixed without rethinking the API. For this reason, I would not expect the issue described in this thread to be resolved anytime soon.

**Comment by mirh** [comment]
Uh, damn. And I suppose it's not possible to just report N/A, null or 0 in those fields? Or somehow ASIO just be a single one entry, which then triggers the rest of its own enumeration only when specifically selected? 
(like.. surely portaudio has its own way to add or remove devices in the middle of execution?)

Otherwise I mean, I don't see why the simple `start_asio` idea (an extra property that you have to toggle on at runtime to actually have ASIO to initialize at all) couldn't work. 
(putting aside that the old as the hills excuse for not having lazy discovery, seems to come from a time when apis were legitimately talking straight with the hardware which I feel should merit a modern round of inquiry from the upper echelons here)

**Comment by dechamps** [comment]
> (like.. surely portaudio has its own way to add or remove devices in the middle of execution?)

No it does not.

**Comment by mirh** [comment]
[Holy](https://github.com/PortAudio/portaudio/issues/11) [moly](https://github.com/PortAudio/portaudio/issues/143)...
Well then, so I suppose the only thing doable for the time being is guarding `PaAsio_Initialize`?
Or is there also some wacky reason that you can't read/accept parameters until devices are enumerated?
